### PR TITLE
Add OC.HealthChecks.Abstractions, and health Checks options to show details

### PR DESCRIPTION
**Pull request by hishamco**
As a part of #13588, moving `HealthChecksOptions` to `OC.HealthChecks.Abstractions` is convenient to allow the health check URL accessible across health checks that are live in other modules

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:40]: Same comment as https://github.com/OrchardCMS/OrchardCore/pull/13592#discussion_r1174454987

It does not do exactly the same, `MapHealthChecks()` will build a new app builder (we have one here) to create a new pipeline (will shortcut the tenant pipeline), so I think that still using `UseHealthChecks()` is better.

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:27]: For info `AddHealthChecks()` registers `HealthCheckPublisherHostedService`, a hosted service that will never get called as we are at the tenant level, not so important but maybe better to just register the `DefaultHealthCheckService` singleton.

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:54]: Idem

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:27]: > For info AddHealthChecks() registers HealthCheckPublisherHostedService

Right, but it doesn't work properly AFAIK, I tried many times, at the end I invoked `IHealthCheckService.Check.CheckHealthAsync()` to enforce register HC services to be triggered

> not so important but maybe better to just register the DefaultHealthCheckService singleton.

I might add it as an extension method on `OC.HealthAbstractions` while hosted service will no longer used

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:27]: > Right, but it doesn't work properly AFAIK

This what I said, a hosted service registered at the tenant level will never get called.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:27]: Unfortunately `DefaultHealthCheckService` is an internal type :(, so I think it's fine to keep it as its

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:27]: Okay then

**Comment by hishamco** [comment]
@sebastienros are we fine with this, so I might need to revisit the other Health Checks PRs

File: src/OrchardCore.Cms.Web/appsettings.json
Comment (Piedone) [inline src/OrchardCore.Cms.Web/appsettings.json:203]: Document this under https://docs.orchardcore.net/en/latest/docs/reference/modules/HealthChecks/.

File: src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:14]:
```suggestion
Gets or sets a value indicating whether to show the details for the checks dependency or not. Defaults to <see langword="false"/>.
```

Comment (Piedone) [inline src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:14]: Explain in the comment what "details" are.

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:49]:
```suggestion
                    ResponseWriter = healthChecksResponseWriter.WriteResponseAsync
```

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Why override the default?

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Models/HealthCheckResponse.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Models/HealthCheckResponse.cs:9]:
```suggestion
public class HealthCheckResponse
```

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Models/HealthCheckEntry.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.HealthChecks/Models/HealthCheckEntry.cs:4]: These comments, and the ones on `HealthCheckReponse`, just repeat what's in the code. This is not very useful. Instead, add further details in the comment, examples, an explanation of how they are used in health checks, so one can understand the properties better.

**Review by Piedone — CHANGES_REQUESTED** [review]
Note to self: check `/health/live` in the next round.

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Because `HealthStatus.Degraded` was shows `200` by default

File: src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs]: I just checked both cases with & without show details everything works as expected

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Yeah, but why is that a problem?

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: I think degraded is sort of problematic that need a proper status code

**Comment by hishamco** [comment]
Anything to address?

**Comment by Piedone** [comment]
We have an ongoing conversation, let's finish that :).

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Yeah, but I guess the ASP.NET Core developers intentionally didn't have that either. My problem here is, that the health check endpoint is an API in itself, that communicates about the health of the app. HTTP 5xx means that something is broken on the server for the current request; in a non-healthy state that may be true, but really only when the health check endpoint itself is broken.

So, I'd actually go in the other direction, and always return a 2xx (looking at those status codes, the only one fitting I see is 200, i.e. return 200 for all states). This only means that the health check endpoint works. Then, the response body will tell whether the app is healthy (after all, this is why we have a body) or not. If there's anything else, than that means that health check is also broken, and the app is definitely not healthy.

Please research whether this has any drawbacks why we shouldn't do it.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: 503 "Service Unavailable" makes sense to me because the specific service is broken, I don't think 200 suits all. We could use 200 for degraded as shown in the [link](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-8.0#customize-the-http-status-code)

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: But the service is not necessarily broken. There can be very well cases where the health check works, but the rest of the site is down. Then the health check can correctly return the status (which is degraded/unhealthy), with itself working, with a 200. Then any non-2xx status could would mean that health check itself is broken too.

How would you otherwise distinguish between health check being broken and the rest of the app working, vs the app being broken and health check working? Because it can also very well be that the app is healthy, but we messed up something with the health check providers.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: This isn't very clear :) assume X mail service is broken, we know the app is healthy but the service itself is broken, so the response above is related to the health check itself, not the app

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: In that case, the response body of the health check will elaborate that Email is broken. I don't see why responding with an HTTP 500 in addition to that would help.

The only use-case I can think is when you ping the health check endpoint with some tool that can't read the response but can only check for an HTTP 200 response code. However, that's really a shortcoming of that tool, then, and you should use something more capable for health checks (and only use such simple pings for checking e.g. the homepage).

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: If you are against the health result status codes, I suggest you open an issue in the AspNetCore repo, and then the team will answer why they chose these codes

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: I'm not against having that capability. But the ASP.NET team gave us the option to specify which status code to use, what you also did here. So, we can make our own decisions on those status codes, just as you did. I'm arguing for using a different configuration than what you opted for.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: So, let rely on the defaults then we can extend and make these codes configurable

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: I've presented my arguments above. I really don't want to have to repeat those at a later time.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: I know, but I'm not convinced with your arguments :)

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: I welcome you to please present your counter-arguments, then.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: > but really only when the health check endpoint itself is broken

Not necessary, if you look to https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503 indicates the service is not ready to handle the request, which is true in terms of health check service, so the service is unhealthy and might require you to retry later

While the health check reports the health of app infrastructure components, how unhealthy components should return 200 if something goes wrong behind the scenes, I know someone will tell me the endpoint is not broken but in reality it's unavailable IMHO

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Would you say that if the health check endpoint is fully working and returns a correct JSON that describes the app otherwise being down, that then the endpoint itself is "Unavailable"?

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Take a GitHub as an example https://www.githubstatus.com/ if `Codespaces` for instance is down for any reason this component is `Unhealthy`, so in terms of health checks the entire app is unhealthy but that doesn't mean  the whole app is down

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs:48]: Yep, and that page follows what I think is a better approach: If Codespaces is down, or even if the whole of GitHub is down, https://www.githubstatus.com/ will still return HTTP 200, and report issues with the services. The only case when https://www.githubstatus.com/ will return 5xx is when https://www.githubstatus.com/ itself (not the rest of GitHub) is broken. The health check endpoint is the same thing, just a simpler approach than a separate status page.

Note that I'm not talking about the Health/Unhealthy flag; I don't have any issues with what's in this PR for that (and I agree that "if Codespaces is down then GitHub as a whole is Unhealthy"). Rather, the only thing I'm talking about is the HTTP status code of the health check endpoint (i.e. what I added the PR comment on above), which I think should be 200 even if the service is unhealthy.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: I revised the PR and the ASP.NET Core docs, both `Healthy` and `Degraded` should return 200, but for `Unhealthy` 503 is the better option I found and agree with. Service is Unavailable makes sense

So let us merge this, and react to the feedback at the next meeting

**Comment by hishamco** [comment]
After a few more tests, I notice:

- `RedisDatabaseFactory` crash if an exception occurs
- `Startup.Order` is no longer needed, I tried some corner cases and everything works as expected. Maybe it was previously for a reason when the code written for OCC

File: src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: I elaborated why it's not better and I'm awaiting your counter arguments on my last comment. There's no point in not settling this now and I absolutely don't want either of us to have to get back to this and kicking the can down the road.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: Also, I elaborate as much as I can on why unhealthy service should return 503 not 200 because the service is unavailable for some reason, I'm not sure why a useless or damaged service should return 200, that's why the default

Let's rely on the default status or hear the feedback from other guys if you never mind, so that PR will not take much time

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: Last example that might change your mind :)

If we have 3 health check services, according to your assumption if Service 1 is unhealthy the status will be 200 for the health reporter, as a user will be happy because the health check reported is 200 while it's not.

If there's no "ShowDetails" option that I have added is there you can recognize if there's an unhealthy service, that's why each health check service should respond either 200 or 503 to notify the reporter if the entire system is healthy or not. Of course, the app SHOULD return the status based on all the health check services

I know this takes so long but I tried in different ways to convince you that we should differentiate between the end-point existence and health check status

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: Is there really a use case when we don't want the health check endpoint to respond with details, and the different services to not provide details when they report themselves being unhealthy?

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: Nope, but responding with 200 does not make sense to me even if you said the API hits :) but that does not mean the response is OK

Let me give you a last example assume you have an API to create a post, for any reason during the post creating a system throws an expectation SHOULD the API respond with 200 or 500?

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: We shouldn't allow empty health check responses, then, and thus remove configuration that we want to discourage (`ShowDetails`).

It should respond with 500. But this is precisely my point :). If the API for creating a blog post is broken, then it should indicate that with a 5xx status code. But you wouldn't want the blog post API return a 500 if only creating a page is broken, right? Only when you want to create a page. Because the page API being broken doesn't mean that the blog API is also broken.

If the health check endpoint is broken then it should return a 5xx too. However, I'm talking about the cases when the health check endpoint itself is working fine, but the rest of the app is unhealthy (then it should still stay 200).

We're probably taking waaaaay too much on this issue, but let me illustrate it with something else. Suppose you see this sign at the side of the road:

image

The sign is working perfectly, but it's telling you about heat and fire. Looks good, right? :) To me, returning 500 is akin to making sure this sign is on fire too, or at least it says "I'm also on fire!". But we don't do that, because we can (and should) distinguish between the sign being broken, and the thing that the sign tells about being broken, because those are two different things.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: > We shouldn't allow empty health check responses

Please elaborate.

> However, I'm talking about the cases when the health check endpoint itself is working fine, but the rest of the app is unhealthy (then it should still stay 200).

I don't think so :) because it tells me that the entire app is healthy, which is not. For that, I need anyone else on the team to share his opinion, maybe I'm wrong

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: > We shouldn't allow empty health check responses

[You said so](https://github.com/OrchardCMS/OrchardCore/pull/13590#discussion_r1565006360) :).

Why would it say the entire app is healthy? The health check will literally return `"unhealthy"`. HTTP 200 only means that this response was successful.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs]: No, not at all, this is not my point. It's that the status of the health check API, the thing that tells about the state of the app, is not the same as the status of the app. The road sign can be fine when it tells that there's a fire ahead. Or githubstatus.com can be fine while github.com, the service it reports about, is broken.

**Comment by hishamco** [comment]
@Piedone if it's fine now let us merge, so this should not take so long. As you know your latest commit will use Fire & Forget to almost all of my PRs, so I need a way to survive :)

File: src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: So didn't [we discuss](https://github.com/OrchardCMS/OrchardCore/pull/13590#discussion_r1565006360) that having a health check response without details doesn't make sense? Because then we don't need this config, and the feature should work as if it were permanently `true`.

File: src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:4]:
```suggestion
/// Represents options for health checks.
```

File: src/OrchardCore.Cms.Web/appsettings.json
Comment (Piedone) [inline src/OrchardCore.Cms.Web/appsettings.json:207]:
```suggestion
    //  "ShowDetails": true
```

File: src/docs/reference/modules/HealthChecks/README.md
Comment (Piedone) [inline src/docs/reference/modules/HealthChecks/README.md:15]: Since `ShowDetails` is not `true` by default:

```suggestion
The following configurations are available and can be customized:
```

File: src/docs/reference/modules/HealthChecks/README.md
Comment (Piedone) [inline src/docs/reference/modules/HealthChecks/README.md:1]: BTW shouldn't the module's display name, and thus the title here as well, be "Health Check**s**"?

File: src/docs/reference/modules/HealthChecks/README.md
Comment (Piedone) [inline src/docs/reference/modules/HealthChecks/README.md:28]:
```suggestion
| `Url` | The relative URL of the health checks endpoint |
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: By default, as we are doing today, hitting the health check the app status, the newly added flag is used if you want to explore all the health checks in the system with their statuses

Comment (Piedone) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: I understand what this does, but not why we'd need it (that it didn't do this before is not a reason in itself).

Comment (hishamco) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: We didn't have any health checks before, but now we do, that's useful to show all the health check details either as JSON or probably later from UI

Comment (Piedone) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: Yes, I'm clear on that. Why do we need the `ShowDetails` option though, when we previously agreed that it doesn't make sense to have it `false`?

Comment (hishamco) [reply src/OrchardCore/OrchardCore.HealthChecks.Abstractions/HealthChecksOptions.cs:16]: Got it that why you pushing to use 200 :)

File: src/OrchardCore.Cms.Web/appsettings.json
Comment (Piedone) [inline src/OrchardCore.Cms.Web/appsettings.json]: After sleeping on it, we do need a config for `ShowDetails`. If a health checks endpoint is publicly available (and given that this is the default and easiest, I believe most are), then one can involuntarily disclose too much information about the app (like the fact that it uses Redis behind the scenes). This is a security issue, while simply returning Healthy/Unhealthy much less so.

So, please revert this change, you were right.

**Review by Piedone — CHANGES_REQUESTED** [review]
Do you still think this needs triage?

File: src/OrchardCore.Cms.Web/appsettings.json
Comment (hishamco) [reply src/OrchardCore.Cms.Web/appsettings.json]: > Do you still think this needs triage?

Nope

**Review by Piedone — APPROVED** [review]

**Comment by Piedone** [comment]
Please check the recording of today's meeting once it's up when we talk about this merge commit (around the 45-minute mark). We might need to reverse course here, due to reinventing the wheel.

**Comment by hishamco** [comment]
> around the 45-minute mark

Really, which reinventing do you mean? I might need to wait until the recording uploaded in YouTube

**Comment by Piedone** [comment]
Yeah, better to watch it.

**Comment by hishamco** [comment]
Sure, I will react to the feedback then

**Comment by Piedone** [comment]
See in the meeting recording [from here](https://youtu.be/CtxQO7tucOc?si=CcR2Fu-8xQeiO4hc&t=3317).

Sebastien talked about [`IHealthCheckPublisher`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.diagnostics.healthchecks.ihealthcheckpublisher?view=dotnet-plat-ext-8.0) seeming to serve the same purpose as our `IHealthChecksResponseWriter`. However, it's not really the same: ours is about creating the health checks HTTP response, not about adding the the `HealthReport`. So, nothing to change.

**Comment by hishamco** [comment]
I will have a look at the recording

**Pull request merged**
